In [3]:
import pandas as pd
import statistics 
import math
from scipy import stats

In [4]:
df_iris=pd.read_csv(r'./iris.data')
df_iris.columns =['sepal_len', 'sepal_width', 'petal_len', 'petal_width','label'] 

In [5]:
df_iris

sepal_len  sepal_width  petal_len  petal_width           label
0          4.9          3.0        1.4          0.2     Iris-setosa
1          4.7          3.2        1.3          0.2     Iris-setosa
2          4.6          3.1        1.5          0.2     Iris-setosa
3          5.0          3.6        1.4          0.2     Iris-setosa
4          5.4          3.9        1.7          0.4     Iris-setosa
..         ...          ...        ...          ...             ...
144        6.7          3.0        5.2          2.3  Iris-virginica
145        6.3          2.5        5.0          1.9  Iris-virginica
146        6.5          3.0        5.2          2.0  Iris-virginica
147        6.2          3.4        5.4          2.3  Iris-virginica
148        5.9          3.0        5.1          1.8  Iris-virginica

[149 rows x 5 columns]

In [6]:
df_setosa = df_iris[df_iris["label"]=="Iris-setosa"]

df_setosa

sepal_len  sepal_width  petal_len  petal_width        label
0         4.9          3.0        1.4          0.2  Iris-setosa
1         4.7          3.2        1.3          0.2  Iris-setosa
2         4.6          3.1        1.5          0.2  Iris-setosa
3         5.0          3.6        1.4          0.2  Iris-setosa
4         5.4          3.9        1.7          0.4  Iris-setosa
5         4.6          3.4        1.4          0.3  Iris-setosa
6         5.0          3.4        1.5          0.2  Iris-setosa
7         4.4          2.9        1.4          0.2  Iris-setosa
8         4.9          3.1        1.5          0.1  Iris-setosa
9         5.4          3.7        1.5          0.2  Iris-setosa
10        4.8          3.4        1.6          0.2  Iris-setosa
11        4.8          3.0        1.4          0.1  Iris-setosa
12        4.3          3.0        1.1          0.1  Iris-setosa
13        5.8          4.0        1.2          0.2  Iris-setosa
14        5.7          4.4        1.5          0.4  Iris-setosa
15        5.4          3.9        1.3          0.4  Iris-setosa
16        5.1          3.5        1.4          0.3  Iris-setosa
17        5.7          3.8        1.7          0.3  Iris-setosa
18        5.1          3.8        1.5          0.3  Iris-setosa
19        5.4          3.4        1.7          0.2  Iris-setosa
20        5.1          3.7        1.5          0.4  Iris-setosa
21        4.6          3.6        1.0          0.2  Iris-setosa
22        5.1          3.3        1.7          0.5  Iris-setosa
23        4.8          3.4        1.9          0.2  Iris-setosa
24        5.0          3.0        1.6          0.2  Iris-setosa
25        5.0          3.4        1.6          0.4  Iris-setosa
26        5.2          3.5        1.5          0.2  Iris-setosa
27        5.2          3.4        1.4          0.2  Iris-setosa
28        4.7          3.2        1.6          0.2  Iris-setosa
29        4.8          3.1        1.6          0.2  Iris-setosa
30        5.4          3.4        1.5          0.4  Iris-setosa
31        5.2          4.1        1.5          0.1  Iris-setosa
32        5.5          4.2        1.4          0.2  Iris-setosa
33        4.9          3.1        1.5          0.1  Iris-setosa
34        5.0          3.2        1.2          0.2  Iris-setosa
35        5.5          3.5        1.3          0.2  Iris-setosa
36        4.9          3.1        1.5          0.1  Iris-setosa
37        4.4          3.0        1.3          0.2  Iris-setosa
38        5.1          3.4        1.5          0.2  Iris-setosa
39        5.0          3.5        1.3          0.3  Iris-setosa
40        4.5          2.3        1.3          0.3  Iris-setosa
41        4.4          3.2        1.3          0.2  Iris-setosa
42        5.0          3.5        1.6          0.6  Iris-setosa
43        5.1          3.8        1.9          0.4  Iris-setosa
44        4.8          3.0        1.4          0.3  Iris-setosa
45        5.1          3.8        1.6          0.2  Iris-setosa
46        4.6          3.2        1.4          0.2  Iris-setosa
47        5.3          3.7        1.5          0.2  Iris-setosa
48        5.0          3.3        1.4          0.2  Iris-setosa

In [7]:
df_versicolor = df_iris[df_iris["label"]=="Iris-versicolor"]

df_versicolor

sepal_len  sepal_width  petal_len  petal_width            label
49        7.0          3.2        4.7          1.4  Iris-versicolor
50        6.4          3.2        4.5          1.5  Iris-versicolor
51        6.9          3.1        4.9          1.5  Iris-versicolor
52        5.5          2.3        4.0          1.3  Iris-versicolor
53        6.5          2.8        4.6          1.5  Iris-versicolor
54        5.7          2.8        4.5          1.3  Iris-versicolor
55        6.3          3.3        4.7          1.6  Iris-versicolor
56        4.9          2.4        3.3          1.0  Iris-versicolor
57        6.6          2.9        4.6          1.3  Iris-versicolor
58        5.2          2.7        3.9          1.4  Iris-versicolor
59        5.0          2.0        3.5          1.0  Iris-versicolor
60        5.9          3.0        4.2          1.5  Iris-versicolor
61        6.0          2.2        4.0          1.0  Iris-versicolor
62        6.1          2.9        4.7          1.4  Iris-versicolor
63        5.6          2.9        3.6          1.3  Iris-versicolor
64        6.7          3.1        4.4          1.4  Iris-versicolor
65        5.6          3.0        4.5          1.5  Iris-versicolor
66        5.8          2.7        4.1          1.0  Iris-versicolor
67        6.2          2.2        4.5          1.5  Iris-versicolor
68        5.6          2.5        3.9          1.1  Iris-versicolor
69        5.9          3.2        4.8          1.8  Iris-versicolor
70        6.1          2.8        4.0          1.3  Iris-versicolor
71        6.3          2.5        4.9          1.5  Iris-versicolor
72        6.1          2.8        4.7          1.2  Iris-versicolor
73        6.4          2.9        4.3          1.3  Iris-versicolor
74        6.6          3.0        4.4          1.4  Iris-versicolor
75        6.8          2.8        4.8          1.4  Iris-versicolor
76        6.7          3.0        5.0          1.7  Iris-versicolor
77        6.0          2.9        4.5          1.5  Iris-versicolor
78        5.7          2.6        3.5          1.0  Iris-versicolor
79        5.5          2.4        3.8          1.1  Iris-versicolor
80        5.5          2.4        3.7          1.0  Iris-versicolor
81        5.8          2.7        3.9          1.2  Iris-versicolor
82        6.0          2.7        5.1          1.6  Iris-versicolor
83        5.4          3.0        4.5          1.5  Iris-versicolor
84        6.0          3.4        4.5          1.6  Iris-versicolor
85        6.7          3.1        4.7          1.5  Iris-versicolor
86        6.3          2.3        4.4          1.3  Iris-versicolor
87        5.6          3.0        4.1          1.3  Iris-versicolor
88        5.5          2.5        4.0          1.3  Iris-versicolor
89        5.5          2.6        4.4          1.2  Iris-versicolor
90        6.1          3.0        4.6          1.4  Iris-versicolor
91        5.8          2.6        4.0          1.2  Iris-versicolor
92        5.0          2.3        3.3          1.0  Iris-versicolor
93        5.6          2.7        4.2          1.3  Iris-versicolor
94        5.7          3.0        4.2          1.2  Iris-versicolor
95        5.7          2.9        4.2          1.3  Iris-versicolor
96        6.2          2.9        4.3          1.3  Iris-versicolor
97        5.1          2.5        3.0          1.1  Iris-versicolor
98        5.7          2.8        4.1          1.3  Iris-versicolor

In [8]:
df_virginica = df_iris[df_iris["label"]=="Iris-virginica"]

df_virginica

sepal_len  sepal_width  petal_len  petal_width           label
99         6.3          3.3        6.0          2.5  Iris-virginica
100        5.8          2.7        5.1          1.9  Iris-virginica
101        7.1          3.0        5.9          2.1  Iris-virginica
102        6.3          2.9        5.6          1.8  Iris-virginica
103        6.5          3.0        5.8          2.2  Iris-virginica
104        7.6          3.0        6.6          2.1  Iris-virginica
105        4.9          2.5        4.5          1.7  Iris-virginica
106        7.3          2.9        6.3          1.8  Iris-virginica
107        6.7          2.5        5.8          1.8  Iris-virginica
108        7.2          3.6        6.1          2.5  Iris-virginica
109        6.5          3.2        5.1          2.0  Iris-virginica
110        6.4          2.7        5.3          1.9  Iris-virginica
111        6.8          3.0        5.5          2.1  Iris-virginica
112        5.7          2.5        5.0          2.0  Iris-virginica
113        5.8          2.8        5.1          2.4  Iris-virginica
114        6.4          3.2        5.3          2.3  Iris-virginica
115        6.5          3.0        5.5          1.8  Iris-virginica
116        7.7          3.8        6.7          2.2  Iris-virginica
117        7.7          2.6        6.9          2.3  Iris-virginica
118        6.0          2.2        5.0          1.5  Iris-virginica
119        6.9          3.2        5.7          2.3  Iris-virginica
120        5.6          2.8        4.9          2.0  Iris-virginica
121        7.7          2.8        6.7          2.0  Iris-virginica
122        6.3          2.7        4.9          1.8  Iris-virginica
123        6.7          3.3        5.7          2.1  Iris-virginica
124        7.2          3.2        6.0          1.8  Iris-virginica
125        6.2          2.8        4.8          1.8  Iris-virginica
126        6.1          3.0        4.9          1.8  Iris-virginica
127        6.4          2.8        5.6          2.1  Iris-virginica
128        7.2          3.0        5.8          1.6  Iris-virginica
129        7.4          2.8        6.1          1.9  Iris-virginica
130        7.9          3.8        6.4          2.0  Iris-virginica
131        6.4          2.8        5.6          2.2  Iris-virginica
132        6.3          2.8        5.1          1.5  Iris-virginica
133        6.1          2.6        5.6          1.4  Iris-virginica
134        7.7          3.0        6.1          2.3  Iris-virginica
135        6.3          3.4        5.6          2.4  Iris-virginica
136        6.4          3.1        5.5          1.8  Iris-virginica
137        6.0          3.0        4.8          1.8  Iris-virginica
138        6.9          3.1        5.4          2.1  Iris-virginica
139        6.7          3.1        5.6          2.4  Iris-virginica
140        6.9          3.1        5.1          2.3  Iris-virginica
141        5.8          2.7        5.1          1.9  Iris-virginica
142        6.8          3.2        5.9          2.3  Iris-virginica
143        6.7          3.3        5.7          2.5  Iris-virginica
144        6.7          3.0        5.2          2.3  Iris-virginica
145        6.3          2.5        5.0          1.9  Iris-virginica
146        6.5          3.0        5.2          2.0  Iris-virginica
147        6.2          3.4        5.4          2.3  Iris-virginica
148        5.9          3.0        5.1          1.8  Iris-virginica

### a. Signal-to-noise ratio 

In [14]:
def calculate_SNR(m_a, m_b, m_c, sd_a, sd_b, sd_c):
    return ((abs(m_a - m_b - m_c)) / (sd_a + sd_b + sd_c))

In [15]:
m_setosa_s_len, m_setosa_s_width, m_setosa_p_len, m_setosa_p_width = df_setosa.mean()
sd_setosa_s_len, sd_setosa_s_width, sd_setosa_p_len, sd_setosa_p_width = df_setosa.std()

m_versi_s_len, m_versi_s_width, m_versi_p_len, m_versi_p_width = df_versicolor.mean()
sd_versi_s_len, sd_versi_s_width, sd_versi_p_len, sd_versi_p_width = df_versicolor.std()

m_vg_s_len, m_vg_s_width, m_vg_p_len, m_vg_p_width = df_virginica.mean()
sd_vg_s_len, sd_vg_s_width, sd_vg_p_len, sd_vg_p_width = df_virginica.std()



In [16]:
m_setosa_s_len, m_setosa_s_width, m_setosa_p_len, m_setosa_p_width

(5.004081632653061, 3.416326530612245, 1.4653061224489796, 0.2448979591836734)

In [17]:
SNR_s_len = calculate_SNR(m_setosa_s_len, m_versi_s_len, m_vg_s_len, sd_setosa_s_len, sd_versi_s_len
                           , sd_vg_s_len)
SNR_s_len

4.986916556479193

In [18]:
SNR_s_width = calculate_SNR(m_setosa_s_width, m_versi_s_width, m_vg_s_width, sd_setosa_s_width, sd_versi_s_width, sd_vg_s_width)
SNR_s_width

2.2796141708770623

In [19]:
SNR_p_len = calculate_SNR(m_setosa_p_len, m_versi_p_len, m_vg_p_len, sd_setosa_p_len, sd_versi_p_len, sd_vg_p_len)
SNR_p_len

6.973789113754186

In [20]:
SNR_p_width = calculate_SNR(m_setosa_p_width, m_versi_p_width, m_vg_p_width, sd_setosa_p_width, sd_versi_p_width, 
                            sd_vg_p_width)
SNR_p_width

5.3521530671583655

Since signal-to-ratio of "petal length" attribute is larger, "Petal length" is a good feature for classification. "Petal width" has the second largest signal-to-noise ratio. Thus petal length and width are the two best features for classification.   

### Two tailed t-test

In [21]:
degree_of_freedom = len(df_setosa) + len(df_versicolor) + len(df_virginica) - 2
degree_of_freedom

147

In [22]:
def calculate_t(inv_cls_cnt, m_a, m_b, m_c, sd_a, sd_b, sd_c):
    return ((abs(m_a - m_b - m_c)) / ((sd_a + sd_b + sd_c) * math.sqrt(inv_class_cnt)))

In [23]:
p = 0.05 #since ocnfidence is 95%
inv_class_cnt = (1/len(df_setosa)) + (1/len(df_versicolor)) + (1/len(df_virginica))
inv_class_cnt

0.060408163265306125

From t-table two tailed, for p=0.05 and degree_of_freedom = 120; t_threshold = 1.980

In [24]:
t_sepal_len = calculate_t(inv_class_cnt, m_setosa_s_len, m_versi_s_len, m_vg_s_len, sd_setosa_s_len, 
                          sd_versi_s_len, sd_vg_s_len)
t_sepal_len

20.290104600610757

In [25]:
t_sepal_width = calculate_t(inv_class_cnt, m_setosa_s_width, m_versi_s_width, m_vg_s_width, 
                            sd_setosa_s_width, sd_versi_s_width, sd_vg_s_width)
t_sepal_width

9.274991761399274

In [26]:
t_petal_len = calculate_t(inv_class_cnt, m_setosa_p_len, m_versi_p_len, m_vg_p_len, sd_setosa_p_len, 
                          sd_versi_p_len, sd_vg_p_len)
t_petal_len

28.374028114994665

In [27]:
t_petal_width = calculate_t(inv_class_cnt, m_setosa_p_width, m_versi_p_width, m_vg_p_width, sd_setosa_p_width,
                            sd_versi_p_width, sd_vg_p_width)
t_petal_width

21.776130468842748

In [34]:
t2, p2 = stats.ttest_ind(df_setosa['sepal_len'],df_virginica['sepal_len'],equal_var=False)
print("t = " + str(t2))
print("p = " + str(p2))

t = -15.332718547400717
p = 3.7898801888654793e-25


In [35]:
t2, p2 = stats.ttest_ind(df_setosa['sepal_width'],df_virginica['sepal_width'],equal_var=False)
print("t = " + str(t2))
print("p = " + str(p2))

t = 6.192754433488112
p = 1.567912726434057e-08


In [36]:
t2, p2 = stats.ttest_ind(df_setosa['petal_len'],df_virginica['petal_len'],equal_var=False)
print("t = " + str(t2))
print("p = " + str(p2))

t = -49.862990878751084
p = 6.676866395755628e-50


In [37]:
t2, p2 = stats.ttest_ind(df_setosa['petal_width'],df_virginica['petal_width'],equal_var=False)
print("t = " + str(t2))
print("p = " + str(p2))

t = -42.60974408939409
p = 9.6628749702774e-49


### Relief

In [ ]:
def calculate_distance(obj1, obj2):
    print('calculate_distance')
    sum = 0
    for i in range(0, 4):
        sum = sum + pow((obj1[i]*obj2[i]), 2)
    return math.sqrt(sum)


def update_distance(obj1,df_iris):
    print('update_distance')
    distance= []
    for i, row in df_iris.iterrows():
        print(row)
        r = row.values.tolist()
        obj2 = r[:4]
        dist = calculate_distance(obj1,obj2)
        print(dist)
        distance.append(distance)
    df_iris['distance'] = distance
    return df_iris

In [ ]:
obj1 = [4.3,3.0,1.1,0.1,]
df_iris_dist = update_distance(obj1,df_iris)
df_iris_dist